File: py-basemapper/tests/test_providers.py
"""Unit tests for Provider Generator Helpers (US4)."""

import json
import warnings

import pytest

from basemapper.providers import (
    EsriRasterProvider,
    EsriVectorProvider,
    RasterProvider,
    VectorProvider,
)


def test_raster_provider_valid_json():
    url = "https://tile.openstreetmap.org/{z}/{x}/{y}.png"
    style = json.loads(str(RasterProvider(url)))
    assert style["version"] == 8
    assert style["sources"]["raster-source"]["tiles"] == [url]
    assert style["layers"][0]["type"] == "raster"


def test_esri_vector_provider_constructs_url():
    base = "https://example.com/VectorTileServer"
    assert str(EsriVectorProvider(base)) == f"{base}/resources/styles/root.json"


def test_esri_vector_provider_strips_trailing_slash():
    base = "https://example.com/VectorTileServer/"
    result = str(EsriVectorProvider(base))
    assert result == "https://example.com/VectorTileServer/resources/styles/root.json"
    assert "//" not in result.split("://")[1]


def test_vector_provider_with_paint():
    paint = {"fill-color": "#e8e0d8", "line-color": "#aaa", "line-width": 1}
    style = json.loads(str(VectorProvider("https://example.com/{z}/{x}/{y}.mvt", paint=paint)))
    fill_layer = next(l for l in style["layers"] if l["type"] == "fill")
    line_layer = next(l for l in style["layers"] if l["type"] == "line")
    assert fill_layer["paint"]["fill-color"] == "#e8e0d8"
    assert line_layer["paint"]["line-color"] == "#aaa"


def test_vector_provider_default_grey():
    style = json.loads(str(VectorProvider("https://example.com/{z}/{x}/{y}.mvt")))
    fill_layer = next(l for l in style["layers"] if l["type"] == "fill")
    assert "fill-color" in fill_layer["paint"]


def test_vector_provider_unknown_key_warns():
    with warnings.catch_warnings(record=True) as w:
        warnings.simplefilter("always")
        style_str = str(VectorProvider("https://x.com/{z}/{x}/{y}.mvt", paint={"circle-radius": 5}))
        assert any("circle-radius" in str(warning.message) for warning in w)
    style = json.loads(style_str)
    for layer in style["layers"]:
        assert "circle-radius" not in layer.get("paint", {})


def test_esri_raster_provider_valid_json():
    url = "https://example.com/MapServer/tile/{z}/{y}/{x}"
    style = json.loads(str(EsriRasterProvider(url)))
    assert style["version"] == 8
    assert style["sources"]["esri-raster-source"]["tileSize"] == 256
    assert style["sources"]["esri-raster-source"]["tiles"] == [url]

File: r-basemapper/tests/testthat/test-providers.R
test_that("raster_provider returns parseable JSON with correct tile URL", {
  url <- "https://tile.openstreetmap.org/{z}/{x}/{y}.png"
  style <- jsonlite::fromJSON(raster_provider(url), simplifyVector = FALSE)
  expect_equal(style$version, 8L)
  expect_equal(style$sources[["raster-source"]]$tiles[[1]], url)
  expect_equal(style$layers[[1]]$type, "raster")
})

test_that("esri_vector_provider constructs correct URL", {
  base <- "https://example.com/VectorTileServer"
  expect_equal(
    esri_vector_provider(base),
    "https://example.com/VectorTileServer/resources/styles/root.json"
  )
})

test_that("esri_vector_provider strips trailing slash without double-slash", {
  result <- esri_vector_provider("https://example.com/VectorTileServer/")
  expect_equal(result, "https://example.com/VectorTileServer/resources/styles/root.json")
  expect_false(grepl("//resources", result))
})

test_that("vector_provider with paint produces correct layer entries", {
  style <- jsonlite::fromJSON(
    vector_provider("https://x.com/{z}/{x}/{y}.mvt",
                    paint = list("fill-color" = "#e8e0d8", "line-color" = "#aaa")),
    simplifyVector = FALSE
  )
  fill_layer <- Filter(function(l) l$type == "fill", style$layers)[[1]]
  line_layer <- Filter(function(l) l$type == "line", style$layers)[[1]]
  expect_equal(fill_layer$paint[["fill-color"]], "#e8e0d8")
  expect_equal(line_layer$paint[["line-color"]], "#aaa")
})

test_that("vector_provider unknown paint key triggers warning", {
  expect_warning(
    vector_provider("https://x.com/{z}/{x}/{y}.mvt", paint = list("circle-radius" = 5)),
    "circle-radius"
  )
})

test_that("vector_provider with empty paint uses default grey values", {
  style <- jsonlite::fromJSON(
    vector_provider("https://x.com/{z}/{x}/{y}.mvt"),
    simplifyVector = FALSE
  )
  fill_layer <- Filter(function(l) l$type == "fill", style$layers)[[1]]
  expect_false(is.null(fill_layer$paint[["fill-color"]]))
})

test_that("esri_raster_provider returns parseable JSON with tileSize 256", {
  url <- "https://example.com/MapServer/tile/{z}/{y}/{x}"
  style <- jsonlite::fromJSON(esri_raster_provider(url), simplifyVector = FALSE)
  expect_equal(style$version, 8L)
  expect_equal(style$sources[["esri-raster-source"]]$tileSize, 256L)
  expect_equal(style$sources[["esri-raster-source"]]$tiles[[1]], url)
})
